### PR TITLE
Export data

### PR DESCRIPTION
**Pull request by annasirunian**
Done:
- [x] store additional gene and reaction data needed for exporting the results
- [x] create endpoint to return 'csv' file

Todo:
- generate 'xlsx' file
- return archive that includes both 'csv' and 'xlsx'

### CODE REVIEW

File: src/metabolic_ninja/resources.py [hunk 1]
```diff
@@ -185,6 +186,147 @@ def get(self):
         return PRODUCT_LIST
 
 
+class PathwayResource(MethodResource):
+    def get(self, job_id, pathway_id):
```
Comment (Midnighter) [inline src/metabolic_ninja/resources.py]: Since there are multiple results per pathway found, please rename this to something else, for example, `row_id`.

```suggestion
    def get(self, job_id, row_id):
```

File: src/metabolic_ninja/resources.py [hunk 2]
```diff
@@ -185,6 +186,147 @@ def get(self):
...
+            response.headers["Content-Type"] = "text/csv"
+            response.headers["Content-Disposition"] = "attachment"
+            return response
+
+
+def get_tabular_data(pathway):
+    result = []
+    targets = pathway["targets"]
+    manipulations = pathway["manipulations"]
+    if pathway["method"] == "PathwayPredictor+DifferentialFVA":
```
Comment (Midnighter) [inline src/metabolic_ninja/resources.py]: Please create separate functions for each method that are then called by this main function.

File: src/metabolic_ninja/worker/designer.py [hunk 3]
```diff
@@ -79,6 +79,7 @@ def evaluate_diff_fva(designs, pathway, model, method):
     logger.info(
         f"Evaluating {len(designs) - 1} differential FVA surface points."
     )
+    table = designs.nth_panel(0)
```
Comment (Midnighter) [inline src/metabolic_ninja/worker/designer.py]: This should only get the diff FVA results for one point so it needs to be the right "panel" for each `design_result` in the below loop. (Based on discussion there seems to be a cameo bug. I will investigate.)

File: src/metabolic_ninja/worker/designer.py [hunk 4]
```diff
@@ -181,6 +191,25 @@ def evaluate_opt_gene(designs, pathway, model, method):
                     for g in design_result.targets
                     if isinstance(g, cameo.core.target.GeneKnockoutTarget)
                 }
+                gene_targets = {}
+                for t in list(knockouts):
```
Comment (Midnighter) [inline src/metabolic_ninja/worker/designer.py]: It seems superfluous to transform knockouts to a list? Also, please use at least three letter variable names outside of list comprehensions.

```suggestion
                for target in knockouts:
```

File: src/metabolic_ninja/worker/designer.py [hunk 5]
```diff
@@ -181,6 +191,25 @@ def evaluate_opt_gene(designs, pathway, model, method):
...
+                    gene_id = t.id
+                    gene = model.genes.get_by_id(gene_id)
+                    gene_targets[gene_id] = []
+                    for r in gene.reactions:
```
Comment (Midnighter) [inline src/metabolic_ninja/worker/designer.py]: Same as above. Maybe, `reaction_target`?

File: src/metabolic_ninja/worker/designer.py [hunk 6]
```diff
@@ -284,6 +321,28 @@ def evaluate_cofactor_swap(designs, pathway, model, method):
...
+    for t in targets:
+        rxn_id = t.id if knockout else t["id"]
+        rxn = model.reactions.get_by_id(rxn_id)
+        result[rxn_id] = {
+            "name": rxn.name,
+            "subsystem": rxn.subsystem,
+            "gpr": rxn.gene_reaction_rule,
+            "definition_of_stoichiometry": rxn.build_reaction_string(True),
+            "flux_reversal": bool(
+                table[table.reaction == rxn_id].flux_reversal.values[0]
```
Comment (Midnighter) [inline src/metabolic_ninja/worker/designer.py]: You can make use of full [pandas indexing](https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html) here:
```suggestion
                table.at[rxn_id, "flux_reversal"]
```

**Review by Midnighter — CHANGES_REQUESTED** [review]
Some smaller changes needed. Otherwise it collects all the required data :+1:

**Comment by annasirunian** [comment]
Thank you! I pushed updates according to your comments

**Review by Midnighter — APPROVED** [review]
Looks great, I look forward to the zip archives :slightly_smiling_face:

File: src/metabolic_ninja/resources.py [hunk 7]
```diff
@@ -185,6 +186,164 @@ def get(self):
...
+                    404,
+                )
+            result = get_tabular_data(pathway)
+            response = make_response(result, status)
+            response.headers["Content-Type"] = "text/csv"
+            response.headers["Content-Disposition"] = "attachment"
+            return response
+
+
+def get_tabular_data(pathway):
```
Comment (kvikshaug) [inline src/metabolic_ninja/resources.py:233]: This logic isn't specific to the HTTP resources so it doesn't belong here. Same with all the added functions below. You could inline it if it's a one-off thing, but it looks to me like it could make sense to encapsulate this in the data model, for example as methods on the `Job` ORM class.

File: src/metabolic_ninja/resources.py [hunk 8]
```diff
@@ -185,6 +186,164 @@ def get(self):
         return PRODUCT_LIST
 
 
+class JobRowResource(MethodResource):
+    def get(self, job_id, row_id):
```
Comment (kvikshaug) [inline src/metabolic_ninja/resources.py:190]: This endpoint should be documented for the API docs

File: src/metabolic_ninja/resources.py [hunk 9]
```diff
@@ -185,6 +186,164 @@ def get(self):
...
+    for target in manipulations:
+        rxn_id = target["id"]
+        rxn_data = {"reaction_target": rxn_id}
+        swapped_from = target["from"]
+        swapped_to = target["to"]
+        rxn_data["swapped_cofactors"] = "; ".join(
+            [
+                swapped_from[i] + " -> " + swapped_to[i]
+                for i in range(len(swapped_from))
+            ]
```
Comment (lukaw3d) [inline src/metabolic_ninja/resources.py:325]: A bit more idiomatic way to do it:
```py
"; ".join(
    a + " -> " + b
    for (a, b) in zip(swapped_from, swapped_to)
)
```

File: src/metabolic_ninja/resources.py [hunk 10]
```diff
@@ -185,6 +186,164 @@ def get(self):
...
+        )
+        rxn_data["reaction_name"] = targets[rxn_id]["name"]
+        rxn_data["subsystem"] = targets[rxn_id]["subsystem"]
+        rxn_data["gpr"] = targets[rxn_id]["gpr"]
+        rxn_data["definition_of_stoichiometry"] = targets[rxn_id][
+            "definition_of_stoichiometry"
+        ]
+        result.append(rxn_data)
```
Comment (lukaw3d) [inline src/metabolic_ninja/resources.py:333]: And if you like jsons:
```py
rxn_id = target["id"]
swapped_from = target["from"]
swapped_to = target["to"]

rxn_data = {
    "reaction_target": rxn_id
    "swapped_cofactors": "; ".join(
        a + " -> " + b
        for (a, b) in zip(swapped_from, swapped_to)
    ),
    "reaction_name": targets[rxn_id]["name"],
    "subsystem": targets[rxn_id]["subsystem"],
    "gpr": targets[rxn_id]["gpr"],
    "definition_of_stoichiometry": targets[rxn_id][
        "definition_of_stoichiometry"
    ]
}
result.append(rxn_data)
```

File: src/metabolic_ninja/resources.py [hunk 11]
```diff
@@ -185,6 +186,164 @@ def get(self):
...
+        except NoResultFound:
+            return (
+                {"error": f"Cannot find any design job with id {job_id}."},
+                404,
+            )
+        else:
+            if job.is_complete():
+                status = 200
+            else:
+                status = 202
```
Comment (kvikshaug) [inline src/metabolic_ninja/resources.py:210]: Why return anything at all if the job isn't completed?

File: src/metabolic_ninja/resources.py [hunk 12]
```diff
@@ -195,4 +354,7 @@ def register(path, resource):
     docs = FlaskApiSpec(app)
     register("/predictions", PredictionJobsResource)
     register("/predictions/<string:job_id>", PredictionJobResource)
+    register(
+        "/predictions/<string:job_id>/<string:row_id>", JobRowResource
```
Comment (kvikshaug) [inline src/metabolic_ninja/resources.py:358]: I think the term "row" is a bit generic. Prediction is probably better since it's already used. (Not part of the PR, but both the URL and the other resource names are also very confusing to me)

File: src/metabolic_ninja/resources.py (hunk 9 above)
Comment (kvikshaug) [reply src/metabolic_ninja/resources.py:325]: Also please prefer f-string formatting, so: `f"{a} -> {b}"`

**Comment by annasirunian** [comment]
@lukaw3d , @kvikshaug , thanks for your review, next PR will contain appropriate changes
